src/test/pthreads/test_pthreads.rs

```diff
@@ -0,0 +1,489 @@
+        free_type(cond);
+        return Err("error: pthread_cond_init failed".into());
+    }
+
+    Ok(cond)
+}
+
+// Destroy a pthread_cond_t
+fn destroy_cond(cond: *mut libc::pthread_cond_t) {
+    if cond != std::ptr::null_mut() {
+        unsafe { libc::pthread_cond_destroy(cond) };
+        free_type(cond);
+    }
+}
+
+// Initialize an allocated pthread_attr_t
+//
+// Returns error if allocation or pthread_attr_init fails
+fn init_attr() -> Result<*mut libc::pthread_attr_t, String> {
+    let attr = alloc_type::<libc::pthread_attr_t>()?;
+
+    if unsafe { libc::pthread_attr_init(attr) } < 0 {
+        free_type(attr);
+        return Err("error: pthread_attr_init failed".into());
+    }
+
+    Ok(attr)
+}
+
+// Destroy a pthread_attr_t
+fn destroy_attr(attr: *mut libc::pthread_attr_t) {
+    if attr != std::ptr::null_mut() {
+        unsafe { libc::pthread_attr_destroy(attr) };
+        free_type(attr);
+    }
+}
+
+// Allocate memory for a specified type
+//
+// Returns error if allocation failed
+fn alloc_type<T>() -> Result<*mut T, String> {
+    let retval = unsafe { libc::malloc(std::mem::size_of::<T>()) as *mut T };
+    if retval == std::ptr::null_mut() {
+        return Err("error: malloc failed!".into());
+    }
+    Ok(retval)
+}
+
+// Free memory from an allocated type
+fn free_type<T>(ptr: *mut T) {
+    if ptr != std::ptr::null_mut() {
+        unsafe { libc::free(ptr as *mut libc::c_void) };
+    }
+}
+
+// Check for error values returned from pthread test functions
+fn check_pthread_error(err: isize) -> Result<(), String> {
+    match err {
+        SUCCESS => Ok(()),
+        MUTEX_LOCK_FAILED => return Err("pthread_mutex_lock failed".into()),
+        MUTEX_UNLOCK_FAILED => return Err("pthread_mutex_unlock failed".into()),
+        COND_WAIT_FAILED => return Err("pthread_cond_wait failed".into()),
+        COND_BROADCAST_FAILED => return Err("pthread_cond_wait failed".into()),
+        NULL_THREAD_ARG => return Err("null thread argument".into()),
+        _ => return Err("unknown error occurred".into()),
+    }
+}
+
+extern "C" fn make_detached(arg: *mut libc::c_void) -> *mut libc::c_void {
+    let p = arg as *mut i32;
+    unsafe { *p += 1 };
+    arg
+}
+
+fn test_make_detached() -> Result<(), String> {
+    let attr = Attr::new()?;
+
+    if unsafe { libc::pthread_attr_setdetachstate(attr.ptr, libc::PTHREAD_CREATE_DETACHED) } < 0 {
+        return Err("error: pthread_attr_setdetachstate failed".into());
+    }
+
+    let mut thread = libc::pthread_t::default();
+    let mut thread_counter = 0_i32;
+
+    let rv = unsafe {
+        libc::pthread_create(
+            &mut thread,
+            attr.ptr,
+            make_detached,
+            &mut thread_counter as *mut _ as *mut libc::c_void,
+        )
+    };
+    if rv != 0 {
+        return Err("error: pthread_create failed".into());
+    }
+
+    while thread_counter == 0 {
+        std::thread::sleep(std::time::Duration::from_millis(1));
+    }
```

Comment:
If your complaint is that this isn't safe, why do you suggest doing something less safe?

Reply:
In this case we need more unsafe code to work around the safety issue in `make_detached()` (the `i32` is accessed in two threads without any memory fence).

Reply:
If your concern is that it is unsafe, why do something less safe? If we want a memory fence not using `pthreads`, safe Rust has those.

Reply:
A volatile read isn't unsafe if you use it correctly. But sure if you want to use a fence then feel free! I just want to avoid anything that requires `futex()` or similar syscalls since I don't want to use too many syscalls in a single test.

Reply:
> I just want to avoid anything that requires futex() or similar syscalls since I don't want to use too many syscalls in a single test.

I'll look into Rust's fencing primitive more to understand what's going on under the hood. Not sure exactly what syscalls are used.

What leads you to thinking the compiler would optimize away reads?

Reply:
How about using something like an [AtomicI32](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicI32.html)?